src/jit/lowerxarch.cpp

```diff
@@ -187,43 +168,44 @@ void Lowering::LowerBlockStore(GenTreeBlk* blkNode)
         // a code sequence of its choice.
         unsigned helperThreshold = max(INITBLK_STOS_LIMIT, INITBLK_UNROLL_LIMIT);
 
-        // TODO-X86-CQ: Investigate whether a helper call would be beneficial on x86
-        if (!blkNode->OperIs(GT_STORE_DYN_BLK) && size <= helperThreshold)
+        if (!blkNode->OperIs(GT_STORE_DYN_BLK) && (size <= helperThreshold))
         {
             // Always favor unrolling vs rep stos.
-            if (size <= INITBLK_UNROLL_LIMIT && initVal->IsCnsIntOrI())
+            if ((size <= INITBLK_UNROLL_LIMIT) && src->IsCnsIntOrI())
             {
+                blkNode->gtBlkOpKind = GenTreeBlk::BlkOpKindUnroll;
+
                 // The fill value of an initblk is interpreted to hold a
                 // value of (unsigned int8) however a constant of any size
                 // may practically reside on the evaluation stack. So extract
                 // the lower byte out of the initVal constant and replicate
                 // it to a larger constant whose size is sufficient to support
                 // the largest width store of the desired inline expansion.
 
-                ssize_t fill = initVal->gtIntCon.gtIconVal & 0xFF;
+                ssize_t fill = src->AsIntCon()->IconValue() & 0xFF;
+
+                if (fill == 0)
+                {
+                    // If the size is multiple of XMM register size there's no need to load 0 in a GPR,
+                    // codegen will use xorps to generate 0 directly in the temporary XMM register.
+                    if ((size % XMM_REGSIZE_BYTES) == 0)
+                    {
+                        src->SetContained();
+                    }
+                }
 #ifdef _TARGET_AMD64_
-                if (size < REGSIZE_BYTES)
+                else if (size >= REGSIZE_BYTES)
                 {
-                    initVal->gtIntCon.gtIconVal = 0x01010101 * fill;
+                    fill *= 0x0101010101010101LL;
+                    src->gtType = TYP_LONG;
```

Comment:
Also, the type was changed to long even when the fill was 0. That isn't necessary and resulted in `xor rax, rax` rather than `xor eax, eax` being generated in disassembly.

Reply:
There are only textual diffs. There are hacks in the emitter that prevented a 64 bit xor from being emitted in the first place, like here: https://github.com/dotnet/coreclr/blob/088299e2b7f439dfd1f5056fdc291b9d82149fc1/src/jit/emitxarch.cpp#L11360-L11366